ビルド時のエラーを修正

rei01saito opened the pull request:
## Checklist before requesting a review
- [x] CIは通っているか
- [x] 依存するPRは存在しないか ~https://github.com/SongCastle/KoR/pull/34 に依存しています~
- [x] issueのリンクが貼られていること

## Related Issue
<!-- ISSUEのリンク -->
- #24 

## Description
<!-- このPRで行った対応について -->
mysql versionの差異によってbuild時、コンテナ立ち上げ時にエラーが発生していました。
そのため下記の修正を行いました。

- mysql_native_passwordをデフォルトで有効にする
- mysqlのversionを固定
- 使用するユーザーを変更

## Context
<!-- 実装の詳細について -->

## How Has This Been Tested?
<!-- テスト方法について -->
1. `make destroy` を実行
2. `make init` を実行
3. `make serve`を実行
4. backend serverが正常に立ち上がっていることを確認

## Problem
<!-- 問題や懸念点があれば記載 -->

## Snapshots
<!-- スクショ等 -->
<img width="1337" alt="スクリーンショット 2024-09-17 3 05 44" src="https://github.com/user-attachments/assets/8ac92f3a-8578-4e61-8410-f26feb908f36">

rei01saito committed:
nmysql_native_passwordに対応させる

rei01saito committed:
userをroot以外に修正

rei01saito committed:
mysqlのバージョンを固定する

rei01saito committed:
Revert "userをroot以外に修正"

This reverts commit ed5b433fe5e989d125ba5019b544046452d78320.

rei01saito committed:
Revert "Revert "userをroot以外に修正""

This reverts commit bb4a94cab6398f2f1def4aa6f4928570421a8cb2.

rei01saito committed:
コンテナ起動時の設定を修正

rei01saito committed:
makefileを修正

SongCastle commented on compose.yaml:
```diff
@@ -25,7 +26,10 @@ services:
-      - MYSQL_ROOT_PASSWORD_FILE=/run/secrets/db-password
+      MYSQL_ROOT_PASSWORD_FILE: /run/secrets/db-password
+      MYSQL_PASSWORD_FILE: /run/secrets/db-password
+      MYSQL_USER: kor
+      MYSQL_DATABASE: kor
```
こちらの 2 つ environment は、compose.yml 上で定義が必要でしょうか ...？
（≒ db/.env から定義されないでしょうか ...？）

- MYSQL_USER
- MYSQL_DATABASE

SongCastle replied:
あっ、 .env ではなく db/.env であるため、
環境変数のセットアップタイミングがビルドのタイミングに間に合わないのかな ...。

rei01saito replied:
@SongCastle
> こちらの 2 つ environment は、compose.yml 上で定義が必要でしょうか ...？
> （≒ db/.env から定義されないでしょうか ...？） 
これは必要になってくるかなと...！

根本の問題としてはMSQL_USER環境変数にrootを指定できないことでした...。
https://qiita.com/30113011tr/items/a046008fbfc3ba2a89f1

そのためcompose.yamlに別のユーザーを定義してup時にそのユーザーを作成するようにしました。
しかしそれだけだと何も権限がなく、migrate実行時にaccess deniedエラーが出てしまいました。

そのためcompose.yamlにdatabaseを指定してそこにアクセスできるような状態にした、という経緯になります...！

SongCastle replied:
ありがとうございます！

> 根本の問題としてはMSQL_USER環境変数にrootを指定できず、コンテナ立ち上げ時にエラーが発生することでした...。
https://qiita.com/30113011tr/items/a046008fbfc3ba2a89f1
>
> そのため.envのMYSQL_USERには別のユーザーを定義する形にしました。

なるほどですね ...

```diff
- MYSQL_USER=root
+ MYSQL_USER=kor
```

そのため、`MYSQL_USER=kor` としていただいたということですね ...！

SongCastle replied:
> そのため.envのMYSQL_USERには別のユーザーを定義する形にしました。
しかしそれだとそのユーザーがまだ作成されておらず、migrate実行時にaccess deniedエラーが出てしまいました。
>
>そのためcompose.yamlにMYSQL_USERとMYSQL_DATABASEを指定して、そのデータベースにアクセスできるユーザーを作成するようにした、という経緯になります...！

こちらの場合、ユーザーが作成された状況になるように、

https://github.com/SongCastle/KoR/pull/35/files/f3601601ccd6ea009d41bbcab12d354e4f885db2..b60155baa64133fbbd99b8c67fc9881b629176f6

SongCastle replied:
> そのため.envのMYSQL_USERには別のユーザーを定義する形にしました。
しかしそれだとそのユーザーがまだ作成されておらず、migrate実行時にaccess deniedエラーが出てしまいました。
>
>そのためcompose.yamlにMYSQL_USERとMYSQL_DATABASEを指定して、そのデータベースにアクセスできるユーザーを作成するようにした、という経緯になります...！

db コンテナの起動プロセスにて、正当な権限を持ったユーザーが作成された状態を担保するのが、
理想かなと思いました！

compose.yaml と db/.env 双方に環境変数が定義されているのは、
混乱を招くような印象があるためです ...。

勝手ながら、コンテナ起動プロセスにてユーザが確保される状況となるようコミットさせていただきました！

https://github.com/SongCastle/KoR/pull/35/files/f3601601ccd6ea009d41bbcab12d354e4f885db2..b60155baa64133fbbd99b8c67fc9881b629176f6

```sh
$ pwd
/path/to/kor

# DB のデータを削除
$ rm -r db/data

$ make build
$ make up
$ make serve 
```

にて、マイグレーションが通過することを確認しています！

お手数ですがご確認いただけると幸いです 🙇

rei01saito replied:
@SongCastle 
> db コンテナの起動プロセスにて、正当な権限を持ったユーザーが作成された状態を担保するのが、
理想かなと思いました！

なるほど...
docker composeの環境変数設定でそれを担保してしまって.envをなくすのも手かなと思ったのですが、それはどうでしょう...？
compose.yamlにてmysqlの環境変数を適切に設定することでdatabase.shで行っている処理を実現できるかと思いました！

> 勝手ながら、コンテナ起動プロセスにてユーザが確保される状況となるようコミットさせていただきました！

流石です！！
マイグレーション通過すること確認できました！
上記docker composeでの解決策よりこちらのほうがよければapproveしてマージしていただいて大丈夫です！

SongCastle replied:
> docker composeの環境変数設定でそれを担保してしまって.envをなくすのも手かなと思ったのですが、それはどうでしょう...？

.env をなくし、compose.yaml に環境変数の値も書くということですね！

.env を設けていることには狙いが 2 つありました。

1. .env を設けることで、秘匿情報を git の管理対象外にする
2. .env の値が複数サービスで必要となる場合に、その値を集約できる

特に、2 について、 .env を廃止した場合、
環境変数の値を複数箇所で参照できなくなってしまうなと ...。

```yaml
#
# .env あり
# (MYSQL_USER 等の値が変わった時に、.env を変更する必要がある)
#

# .env
MYSQL_USER=kor
MYSQL_DATABASE=kor

---

# yaml
services:
  mysql:
    environment:
      MYSQL_USER: "${MYSQL_USER}"
      MYSQL_DATABASE: "${MYSQL_DATABASE}"

  go:
    environment:
      MYSQL_USER: "${MYSQL_USER}"
      MYSQL_DATABASE: "${MYSQL_DATABASE}" 
```

---

```yaml
#
# .env なし
# (MYSQL_USER 等の値が変わった時に、compose.yaml の各 services の environment を変更する必要がある)
#

# yaml
services:
  mysql:
    environment:
      MYSQL_USER: kor
      MYSQL_DATABASE: kor

  go:
    environment:
      MYSQL_USER: kor
      MYSQL_DATABASE: kor
```

ご見解を伺えると幸いです！！ 🙇

SongCastle replied:
> compose.yamlにてmysqlの環境変数を適切に設定することでdatabase.shで行っている処理を実現できるかと思いました！

MySQL 公式が提供する環境変数だけで完結できないものを含めてしまっているんですよね ...
（`$MYSQL_DATABASE_TEST` : バックエンドの unittest 用に確保していた DB。今は使っていない ... 多分 ...）

ですので、database.sh は `$MYSQL_DATABASE_TEST` のみを setup スクリプトとするのは、
一案かなと思います！

個人的には、 database.sh は アプリケーション稼働に必要な DB の setup スクリプトとする方が、
見通しが良いのかなと考えていました！
※ database.sh のファイル単体で見た人が「ふーん、DB は `$MYSQL_DATABASE_TEST` のみを利用しているのか ... （他の DB は存在しないのかな ...）」とならないように ... というイメージ

rei01saito replied:
@SongCastle 
> .env を設けていることには狙いが 2 つありました。

すみません、どのみちenvは必須でしたね...
僕が気になっていたのはcompose.yamlでデータベース、ユーザーともに作成できるのにわざわざdatabase.shにてその処理を行う必要があるのかという部分でした。

> MySQL 公式が提供する環境変数だけで完結できないものを含めてしまっているんですよね ...
>ですので、database.sh は $MYSQL_DATABASE_TEST のみを setup スクリプトとするのは、
一案かなと思います！

なるほど、これであれば別途スクリプトで作成する必要がありますね...！

> 個人的には、 database.sh は アプリケーション稼働に必要な DB の setup スクリプトとする方が、
見通しが良いのかなと考えていました！

いやどうだろう...
database.shファイルの命名をcreate_test_database.shといったように具体的な処理のものにしてしまって、テストデータベースのみここで作成するほうがより具体的でわかりやすいような...
ちょっとdatabase.shだと名前が抽象的なのと、compose.yamlでデータベース作成することの方が一般的に多いだろうから少し気になってくるかなと。
好みの問題かもしれないけどどうかな？


最終的な判断は任せます！！

rei01saito commented on db/docker-entrypoint-initdb.d/databash.sh:
```diff
@@ -1,10 +1,5 @@
 #!/bin/bash
 
-if [ -z $(which mysql) ]; then
-  echo 'Not found mysql command'
-  exit 1
-fi
```
@SongCastle 
ちなみにこの部分を削除した理由ですが、whichコマンドがないのでここでエラーが発生してしまっていました...

SongCastle replied:
（ベースイメージから which がなくなってしまった、ということですね ...）

SongCastle committed:
[db] compose.yaml の環境変数を db/.env を参照するようにする

db の起動プロセスにて、MYSQL_USER が MYSQL_DATABASE にアクセスできるようにすることで、
compose.yaml での環境変数の定義を削除する。

SongCastle committed:
[db] databash.sh → database.sh へ名前変更する

typo であるため。

SongCastle committed:
[back] migrate.sh で参照するパスワードを MYSQL_USER のものにする

MYSQL_USER としてマイグレーションを実行するため。

SongCastle committed:
[back] README.md のミスを修正する

rei01saito commented:
@SongCastle 
あといくつかissueの作成と既存のものにコメントしているのでお手隙の際にご確認よろしくお願いします～
